File: tools/tracer/src/tracer/commands/symbols.py
"""`trace symbols <file>` — module-level symbols of a file from the architecture graph."""

from __future__ import annotations

import json
import sys
from pathlib import Path

import click

from tracer import architecture, cache
from tracer.deps import require_dependencies


@click.command()
@click.argument("file", type=click.Path(exists=True, dir_okay=False))
@click.option("--json", "as_json", is_flag=True)
def command(file: str, as_json: bool) -> None:
    """Module-level symbols of `file`, sourced from the architecture graph."""
    require_dependencies()
    target = Path(file).resolve()
    repo_root = cache.repo_root_for(target)
    try:
        relative = str(target.relative_to(repo_root.resolve()))
    except ValueError:
        relative = str(target)

    graph = architecture.get(repo_root)
    file_symbols = [
        node for node in graph.nodes.values()
        if node.source_file == relative and node.kind != "module"
    ]

    if as_json:
        click.echo(
            json.dumps(
                {
                    "file": relative,
                    "symbols": [
                        {
                            "node_id": s.id,
                            "label": s.label,
                            "kind": s.kind,
                            "source_line": s.source_line,
                        }
                        for s in file_symbols
                    ],
                    "symbol_count": len(file_symbols),
                },
                indent=2,
            )
        )
        return

    click.echo(f"Symbols in {relative} ({len(file_symbols)}):")
    if not file_symbols:
        click.echo("  (no module-level symbols found in architecture graph)")
        click.echo("  (file may not have been extracted — check supported extensions via `trace doctor`)")
        return
    for s in sorted(file_symbols, key=lambda n: n.source_line or 0):
        click.echo(f"  L{s.source_line:<5} [{s.kind}] {s.label}")

File: packages/claude/bin/tracer-dist/crate/src/commands/symbols.rs
//! `trace symbols <file>` — module-level symbols of a file from the
//! architecture graph.

use crate::{architecture, cache};
use anyhow::Result;
use serde_json::{json, Value};
use std::path::Path;

pub fn run(file: &Path, as_json: bool) -> Result<Value> {
    crate::pathval::require_file(file, "FILE");
    let target = cache::absolutize(file);
    let repo_root = cache::repo_root_for(&target);
    let relative = cache::relative_to_root(&target, &repo_root);

    let graph = architecture::get(&repo_root);
    let mut file_symbols: Vec<&architecture::Node> = graph
        .node_order
        .iter()
        .filter_map(|id| graph.nodes.get(id))
        .filter(|n| n.source_file.as_deref() == Some(relative.as_str()) && n.kind != "module")
        .collect();

    let symbols: Vec<_> = file_symbols
        .iter()
        .map(|s| {
            json!({
                "node_id": s.id,
                "label": s.label,
                "kind": s.kind,
                "source_line": s.source_line,
            })
        })
        .collect();
    let out = json!({
        "file": relative,
        "symbols": symbols,
        "symbol_count": file_symbols.len(),
    });

    if !as_json {
        println!("Symbols in {relative} ({}):", file_symbols.len());
        if file_symbols.is_empty() {
            println!("  (no module-level symbols found in architecture graph)");
            println!("  (file may not have been extracted — check supported extensions via `trace doctor`)");
        } else {
            file_symbols.sort_by_key(|n| n.source_line.unwrap_or(0));
            for s in &file_symbols {
                println!(
                    "  L{:<5} [{}] {}",
                    s.source_line.unwrap_or(0),
                    s.kind,
                    s.label,
                );
            }
        }
    }
    Ok(out)
}
